shards/aggregate.go

```diff
@@ -154,15 +157,30 @@ func newFlushCollectSender(opts *zoekt.SearchOptions, sender zoekt.Sender) (zoek
 	}), finalFlush
 }
 
-// limitSender wraps a sender and calls cancel once it has seen limit file
-// matches.
-func limitSender(cancel context.CancelFunc, sender zoekt.Sender, limit int) zoekt.Sender {
+// limitSender wraps a sender and calls cancel once it has seen either of
+// `docLimit` files or `matchLimit` matches.
+func limitSender(cancel context.CancelFunc, sender zoekt.Sender, docLimit, matchLimit int, chunkMatches bool) zoekt.Sender {
+	docLimited := docLimit > 0
+	matchLimited := matchLimit > 0
 	return stream.SenderFunc(func(result *zoekt.SearchResult) {
-		if len(result.Files) >= limit {
-			result.Files = result.Files[:limit]
-			cancel()
+		// We cancel when we hit either of the two limits. It is assumed that,
+		// after canceling, we are not called again.
```

Comment:
I wonder if this assumption has a lurking race condition we haven't noticed before. Cancel doesn't immedietly stop calls to Send. I'll follow up.